libraries/AP_HAL_Linux/UARTDriver.cpp

```diff
@@ -417,7 +334,7 @@ int UARTDriver::_write_fd(const uint8_t *buf, uint16_t n)
     ret = _device->write(buf, n);
 
     if (ret > 0) {
-        BUF_ADVANCEHEAD(_writebuf, ret);
+        _writebuf.advance(ret);
```

Comment:
> That would be a mistake because it would allow the ringbuffer to overwrite this buffer segment prior consumption by this function, which only happens the line above the if on this function.

The advance would be done after the _write_fd completed.
Again, I understand that you want to change the least possible, but you did change the `_read_fd`.